plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs

```diff
@@ -372,6 +376,82 @@ split (Value mp) = (negate (Value neg), Value pos) where
+   check equality of 'TokenName'-value pairs, since the value of a 'TokenName' is an 'Integer' and
+   @(==) :: Integer -> Integer -> Bool@ is generally much faster than repeatedly searching for keys
+   in a list
+3. having some clever logic for computing equality of values right away in some cases, but not in
+   others would not only complicate the algorithm, but also increase the size of the function and
+   this resource is quite scarce as the size of a program growing beyond what's acceptable by the
+   network can be a real deal breaker, while general performance concerns don't seem to be as
+   pressing
+-}
+unordEqWith :: forall k v. Eq k => (v -> Bool) -> (v -> v -> Bool) -> [(k, v)] -> [(k, v)] -> Bool
+unordEqWith is0 eqV = goBoth where
+    -- Recurse on the spines of both the lists simultaneously.
+    goBoth :: [(k, v)] -> [(k, v)] -> Bool
+    -- One spine is longer than the other one, but this still can result in a succeeding equality
+    -- check if the non-empty list only contains zero values.
+    goBoth []                 kvsR                             = all (is0 . snd) kvsR
+    -- Symmetric to the previous case.
+    goBoth kvsL               []                               = all (is0 . snd) kvsL
+    -- Both spines are non-empty.
+    goBoth ((kL, vL) : kvsL') kvsR0@(kvR0@(kR0, vR0) : kvsR0')
+        -- We could've avoided having this clause if we always searched for the right key-value pair
+        -- using @goRight@, however the sheer act of invoking that function, passing an empty list
+        -- to it as an accumulator and calling 'appendR' afterwards affects performance quite a bit,
+        -- considering that all of that happens for every single element of the left list. Hence we
+        -- handle the special case of lists being equal pointwise (or at least their prefixes being
+        -- equal pointwise) with a bit of additional logic to get some easy performance gains.
+        | kL == kR0  = if vL `eqV` vR0 then goBoth kvsL' kvsR0' else False
+        | is0 vL     = goBoth kvsL' kvsR0
+        | otherwise  = goRight [kvR0 | not $ is0 vR0] kvsR0'
+        where
+            -- Recurse on the spine of the right list looking for a key-value pair whose key matches
+            -- @kL@, i.e. the first key in the remaining part of the left list. The accumulator
+            -- contains (in reverse order) all elements of the right list processed so far whose
+            -- keys are not equal to @kL@ and values are non-zero.
+            goRight :: [(k, v)] -> [(k, v)] -> Bool
+            goRight _   []                     = False
+            goRight acc (kvR@(kR, vR) : kvsR')
+                | is0 vR    = goRight acc kvsR'
+                -- @appendR@ recreates @kvsR0'@ with @(kR, vR)@ removed, since that pair
+                -- equals @(kL, vL)@ from the left list, hence we throw both of them away.
+                | kL == kR  = if vL `eqV` vR then goBoth kvsL' (appendR acc kvsR') else False
```

Comment:
> that's unlikely to make much difference since we'll only reach this code if the two lists already fail to be in the same order. 

Yeah, but what if the difference is only in two elements being swapped or something. I think it'd be too surprising to reverse the order of elements all the time for anybody who tried to understand the performance of `(==)`, so I opted for the predictable version that is likely more efficient as well.

> (Incidentally, I think `appendR` is often called `revAppend` in the strict world).

I'll rename, thanks.